File: 2024/Day 1/d01-hard.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <unordered_map>

int main() {
    std::ifstream inputFile("input.txt");
    if (!inputFile.is_open()) {
        std::cerr << "Error: Could not open the file!" << std::endl;
        return 1;
    }

    std::vector<int> l, r;
    std::string line;

    // Read each line from the file
    while (std::getline(inputFile, line)) {
        std::istringstream lineStream(line);
        int num1, num2;
        
        // Parse two numbers separated by whitespace
        if (lineStream >> num1 >> num2) {
            l.push_back(num1);
            r.push_back(num2);
        }
    }

    inputFile.close();



    // Count occurrences of each number in r
    std::unordered_map<int, int> r_count;
    for (const auto& num : r) {
        r_count[num]++;
    }

    // Calculate the total similarity score
    int s = 0;
    for (const auto& num : l) {
        s += num * r_count[num]; // Multiply the number by its count in r
    }

    std::cout << s << std::endl;

    return 0;

}

File: 2024/Day 1/d01-hard.py
with open('input.txt', 'r') as f:
    lines = f.readlines()
    l,r=[], []
    #each line is a serries of 2 numbers serpated by white spaces
    #example 50472   55227
    # seperate each set into 2 lists, sort those lists so the smallest number of each list is at the top
    for line in lines:
        nums = line.strip().split()
        if len(nums) == 2:  # Ensure there are exactly two numbers
            l.append(int(nums[0]))
            r.append(int(nums[1]))

s=0

#This time, you'll need to figure out exactly how often each number from the left list appears in the right list. Calculate a total similarity score by adding up each number in the left list after multiplying it by the number of times that number appears in the right list.
for i in l:
    s+= i * r.count(i)
print(s)